src/arkode/arkode_arkstep_io.c

```diff
@@ -677,12 +679,66 @@ int arkStep_SetDefaults(ARKodeMem ark_mem)
   step_mem->msbp     = MSBP;  /* max steps between updates to J or P */
   step_mem->stages   = 0;     /* no stages */
   step_mem->istage   = 0;     /* current stage */
-  step_mem->Be       = NULL;  /* no Butcher tables */
-  step_mem->Bi       = NULL;
-  step_mem->NLS      = NULL; /* no nonlinear solver object */
   step_mem->jcur     = SUNFALSE;
   step_mem->convfail = ARK_NO_FAILURES;
   step_mem->stage_predict = NULL; /* no user-supplied stage predictor */
+
+  /* Remove pre-existing Butcher tables */
+  if (step_mem->Be)
+  {
+    ARKodeButcherTable_Space(step_mem->Be, &Bliw, &Blrw);
+    ark_mem->liw -= Bliw;
+    ark_mem->lrw -= Blrw;
+    ARKodeButcherTable_Free(step_mem->Be);
+  }
+  step_mem->Be = NULL;
+  if (step_mem->Bi)
+  {
+    ARKodeButcherTable_Space(step_mem->Bi, &Bliw, &Blrw);
+    ark_mem->liw -= Bliw;
+    ark_mem->lrw -= Blrw;
+    ARKodeButcherTable_Free(step_mem->Bi);
+  }
+  step_mem->Bi = NULL;
+
+  /* Remove pre-existing nonlinear solver object */
+  if (step_mem->NLS && step_mem->ownNLS) { SUNNonlinSolFree(step_mem->NLS); }
+  step_mem->NLS = NULL;
+
+  /* Remove pre-existing SUNAdaptController object, and replace with "PID" */
+  if (ark_mem->hadapt_mem->owncontroller)
+  {
+    retval = SUNAdaptController_Space(ark_mem->hadapt_mem->hcontroller, &lenrw,
+                                      &leniw);
+    if (retval == SUN_SUCCESS)
```

Comment:
Does `SUNAdaptController` return `0` by default? If so, I think this check can be removed.

Reply:
The base `SUNAdaptController` class returns whatever integer return value is sent by the implementation.  Both of the provided `SUNAdaptController` implementations (Soderlind and ImExGus) indeed return `0` no matter what, but users could do differently (the API only states that they should return a `SUNErrCode` indicating success or failure).